platooning_control/src/platoon_control.cpp

```diff
@@ -13,311 +13,427 @@
+    if (tp->trajectory_points.size() < 2) {
+            RCLCPP_WARN_STREAM(get_logger(), "PlatoonControlPlugin cannot execute trajectory as only 1 point was provided");
+            return;
+        }
 
+        latest_trajectory_ = *tp;
+        prev_input_time_ = this->now().nanoseconds() / 1000000;
+        ++consecutive_input_counter_;
+        RCLCPP_DEBUG_STREAM(get_logger(), "New trajectory plan #" << consecutive_input_counter_ << " at time " << prev_input_time_);
+        rclcpp::Time tp_time(tp->header.stamp);
+        RCLCPP_DEBUG_STREAM(get_logger(), "tp header time =                " << tp_time.nanoseconds() / 1000000);
+  }
+
+  geometry_msgs::msg::TwistStamped PlatoonControlPlugin::composeTwistCmd(double linear_vel, double angular_vel)
+  {
+    geometry_msgs::msg::TwistStamped cmd_twist;
+    cmd_twist.twist.linear.x = linear_vel;
+    cmd_twist.twist.angular.z = angular_vel;
+    cmd_twist.header.stamp = this->now();
+    return cmd_twist;
+  }
+
+  autoware_msgs::msg::ControlCommandStamped PlatoonControlPlugin::composeCtrlCmd(double linear_vel, double steering_angle)
+  {
+    autoware_msgs::msg::ControlCommandStamped cmd_ctrl;
+    cmd_ctrl.header.stamp = this->now();
+    cmd_ctrl.cmd.linear_velocity = linear_vel;
+    RCLCPP_DEBUG_STREAM(get_logger(), "ctrl command speed " << cmd_ctrl.cmd.linear_velocity);
+    cmd_ctrl.cmd.steering_angle = steering_angle;
+    RCLCPP_DEBUG_STREAM(get_logger(), "ctrl command steering " << cmd_ctrl.cmd.steering_angle);
+
+    return cmd_ctrl;
+  }
+
+  bool PlatoonControlPlugin::get_availability() {
+    return true; // TODO for user implement actual check on availability if applicable to plugin
+  }
+
+  std::string PlatoonControlPlugin::get_version_id() {
+    return "v1.0";
+  }
+
+  // extract maximum speed of trajectory
+  double PlatoonControlPlugin::getTrajectorySpeed(std::vector<carma_planning_msgs::msg::TrajectoryPlanPoint> trajectory_points)
+  {
+    double trajectory_speed = 0;
+
+    double dx1 = trajectory_points[trajectory_points.size()-1].x - trajectory_points[0].x;
+    double dy1 = trajectory_points[trajectory_points.size()-1].y - trajectory_points[0].y;
+    double d1 = sqrt(dx1*dx1 + dy1*dy1);
+    double t1 = (rclcpp::Time((trajectory_points[trajectory_points.size()-1].target_time)).nanoseconds() - rclcpp::Time(trajectory_points[0].target_time).nanoseconds())/1e9;
+
+    double avg_speed = d1/t1;
+    RCLCPP_DEBUG_STREAM(get_logger(), "trajectory_points size = " << trajectory_points.size() << ", d1 = " << d1 << ", t1 = " << t1 << ", avg_speed = " << avg_speed);
+
+    for(size_t i = 0; i < trajectory_points.size() - 2; i++ )
+    {
+        double dx = trajectory_points[i + 1].x - trajectory_points[i].x;
+        double dy = trajectory_points[i + 1].y - trajectory_points[i].y;
+        double d = sqrt(dx*dx + dy*dy);
+        double t = (rclcpp::Time((trajectory_points[i + 1].target_time)).nanoseconds() - rclcpp::Time(trajectory_points[i].target_time).nanoseconds())/1e9;
```

Comment:
There is a function called `seconds()` which we can directly used instead of doing these conversions